// doge-config/lib/index.js
'use strict';
var __createBinding =
    (this && this.__createBinding) ||
    (Object.create
        ? function (o, m, k, k2) {
              if (k2 === undefined) k2 = k;
              var desc = Object.getOwnPropertyDescriptor(m, k);
              if (
                  !desc ||
                  ('get' in desc
                      ? !m.__esModule
                      : desc.writable || desc.configurable)
              ) {
                  desc = {
                      enumerable: true,
                      get: function () {
                          return m[k];
                      },
                  };
              }
              Object.defineProperty(o, k2, desc);
          }
        : function (o, m, k, k2) {
              if (k2 === undefined) k2 = k;
              o[k2] = m[k];
          });
var __setModuleDefault =
    (this && this.__setModuleDefault) ||
    (Object.create
        ? function (o, v) {
              Object.defineProperty(o, 'default', {
                  enumerable: true,
                  value: v,
              });
          }
        : function (o, v) {
              o['default'] = v;
          });
var __importStar =
    (this && this.__importStar) ||
    function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (
                    k !== 'default' &&
                    Object.prototype.hasOwnProperty.call(mod, k)
                )
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    };
Object.defineProperty(exports, '__esModule', { value: true });
exports.normalizeConfigName =
    exports.getConfigDir =
    exports.getConfig =
    exports.types =
    exports.ConfigMap =
    exports.ConfigField =
    exports.ConfigDir =
    exports.ConfigArray =
    exports.default =
    exports.Config =
        void 0;
var Config_1 = require('./classes/Config');
Object.defineProperty(exports, 'Config', {
    enumerable: true,
    get: function () {
        return Config_1.Config;
    },
});
Object.defineProperty(exports, 'default', {
    enumerable: true,
    get: function () {
        return Config_1.Config;
    },
});
var ConfigArray_1 = require('./classes/ConfigArray');
Object.defineProperty(exports, 'ConfigArray', {
    enumerable: true,
    get: function () {
        return ConfigArray_1.ConfigArray;
    },
});
var ConfigDir_1 = require('./classes/ConfigDir');
Object.defineProperty(exports, 'ConfigDir', {
    enumerable: true,
    get: function () {
        return ConfigDir_1.ConfigDir;
    },
});
var ConfigField_1 = require('./classes/ConfigField');
Object.defineProperty(exports, 'ConfigField', {
    enumerable: true,
    get: function () {
        return ConfigField_1.ConfigField;
    },
});
var ConfigMap_1 = require('./classes/ConfigMap');
Object.defineProperty(exports, 'ConfigMap', {
    enumerable: true,
    get: function () {
        return ConfigMap_1.ConfigMap;
    },
});
exports.types = __importStar(require('./types'));
var getConfig_1 = require('./utils/getConfig');
Object.defineProperty(exports, 'getConfig', {
    enumerable: true,
    get: function () {
        return getConfig_1.getConfig;
    },
});
var getConfigDir_1 = require('./utils/getConfigDir');
Object.defineProperty(exports, 'getConfigDir', {
    enumerable: true,
    get: function () {
        return getConfigDir_1.getConfigDir;
    },
});
var normalizeConfigName_1 = require('./utils/normalizeConfigName');
Object.defineProperty(exports, 'normalizeConfigName', {
    enumerable: true,
    get: function () {
        return normalizeConfigName_1.normalizeConfigName;
    },
});
