// doge-json/lib/index.js
'use strict';
var __importDefault =
	(this && this.__importDefault) ||
	function (mod) {
		return mod && mod.__esModule ? mod : { default: mod };
	};
Object.defineProperty(exports, '__esModule', { value: true });
exports.write =
	exports.read =
	exports.normalize_object =
	exports.fs =
	exports.decode =
	exports.encode =
		void 0;
const decode_1 = __importDefault(require('./decode'));
exports.decode = decode_1.default;
const normalize_and_encode_1 = require('./normalize-and-encode');
Object.defineProperty(exports, 'encode', {
	enumerable: true,
	get: function () {
		return normalize_and_encode_1.encode;
	},
});
Object.defineProperty(exports, 'normalize_object', {
	enumerable: true,
	get: function () {
		return normalize_and_encode_1.normalize_object;
	},
});
const read_1 = __importDefault(require('./read'));
exports.read = read_1.default;
const write_1 = __importDefault(require('./write'));
exports.write = write_1.default;
const fs_polyfill_1 = __importDefault(require('./fs-polyfill'));
exports.fs = fs_polyfill_1.default;
exports.default = {
	encode: normalize_and_encode_1.encode,
	decode: decode_1.default,
	fs: fs_polyfill_1.default,
	normalize_object: normalize_and_encode_1.normalize_object,
	read: read_1.default,
	write: write_1.default,
};
