// ps-std/lib/classes/Queue.js
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.Queue = void 0;
class Queue {
	constructor(error_handler) {
		/** Is a process ongoing? */
		this.working = false;
		/** Queued callbacks */
		this.queue = Array();
		this.error_handler = error_handler;
	}
	/** Call next() after call stack resolved */
	next_async() {
		setTimeout(() => this.next());
	}
	/** Call the next callback */
	next() {
		const next = this.queue.shift();
		try {
			return next ? next() : (this.working = false);
		} catch (error) {
			if (this.error_handler) {
				this.error_handler(error);
			} else {
				throw error;
			}
		}
	}
	/** Add a callback */
	add(cb) {
		if (this.working) {
			return this.queue.push(cb);
		} else {
			this.working = true;
			return cb();
		}
	}
	/** Get a promise */
	get promise() {
		return new Promise((resolve) => this.add(resolve));
	}
}
exports.Queue = Queue;
exports.default = Queue;
Object.defineProperties(Queue, {
	default: { get: () => Queue },
	Queue: { get: () => Queue },
});
